src/fs/types.rs

```diff
@@ -0,0 +1,212 @@
+use std::{hash::{Hash, Hasher}, path::{Path, PathBuf}};
+use smash_arc::{Hash40, Region};
+use std::ops::Deref;
+
+#[derive(Debug)]
+pub enum RejectionReason {
+    DuplicateFile(PathBuf),
+    NotFound(SmashPath),
+    MissingExtension
+}
+
+#[repr(transparent)]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct SmashPath(pub PathBuf);
+
+impl Deref for SmashPath {
+    type Target = Path;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl AsRef<Path> for SmashPath {
+    fn as_ref(&self) -> &Path {
+        &self
+    }
+}
+
+impl From<SmashPath> for PathBuf {
+    fn from(smashpath: SmashPath) -> Self {
+        smashpath.0
+    }
+}
+
+impl SmashPath {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn from<P: AsRef<Path>>(path: P) -> Result<Self, String> {
+        let path = path.as_ref();
+
+        if path.extension().is_some() {
+            Ok(Self(path.to_path_buf()))
+        } else {
+            Err(String::from("This path does not have an extension"))
```

Comment:
Should this be returning a `String` instead of a `RejectionReason` as the error? Maybe this string should be a `fmt::Display` impl for `RejectionReason::MissingExtension` instead?

Reply:
If a file is rejected it will never get this far, since a SmashPath is only created for files that. RejectionReason also only exists for file discovery, and technically SmashPath is not file discovery, so returning it here does not make very much sense.